packages/astro/src/i18n/index.ts

```diff
@@ -298,9 +298,14 @@ export function redirectToDefaultLocale({
 }
 
 // NOTE: public function exported to the users via `astro:i18n` module
-export function notFound({ base, locales }: MiddlewarePayload) {
+export function notFound({ base, locales, fallback }: MiddlewarePayload) {
 	return function (context: APIContext, response?: Response): Response | undefined {
-		if (response?.headers.get(REROUTE_DIRECTIVE_HEADER) === 'no') return response;
+		if (
+			response?.headers.get(REROUTE_DIRECTIVE_HEADER) === 'no' &&
+			typeof fallback === 'undefined'
```

Comment:
Even though we're about to render the internal `404` or `500`, we still want to handle the fallback